Dependency injection support

CodeBlanch opened the pull request:
This has been a hot topic lately so I took a stab at making it work. This isn't done, but enough of it is there to demonstrate the concept. Looking for feedback, please fire away!

## Design Considerations

* ~The main goal is for only OpenTelemetry.Extensions.Hosting to have a dependency on anything in Microsoft.Extensions.* suite.~ Turns out that dependency is already there so the main goal now is to keep .NET Framework users away from having to use IServiceProvider to build TracerProvider. You should only need that if you are using Hosting.

## Public API Changes

### SDK

```csharp
    // TracerProviderBuilderSdk renamed TracerProviderBuilderBase and is now public so
    // Hosting library can use it as a base class. Tried to avoid this but lots of things with logic like:
    //   if (tracerProviderBuilder is TracerProviderBuilderSdk tracerProviderBuilderSdk)
    public abstract class TracerProviderBuilderBase : TracerProviderBuilder
    {
        // Already existed but is now public because the class is public
        public override TracerProviderBuilder AddInstrumentation<TInstrumentation>(
            Func<TInstrumentation> instrumentationFactory)
            where TInstrumentation : class

        // Already existed but is now public because the class is public
        public override TracerProviderBuilder AddSource(params string[] names)

        // TracerProviderBuilderHosting needs a way to register instrumentation without the generic
        protected TracerProviderBuilder AddInstrumentation(
            string instrumentationName,
            string instrumentationVersion,
            Func<object> instrumentationFactory)

        // Needed so hosting builder can call the base builder without using the extension which has a check
        // preventing it from executing for IDeferredTracerProviderBuilders
        protected TracerProvider Build()
    }

    // A new interface to enable libraries with an SDK-dependency to perform 
    // late-bound TracerProviderBuilder configuration once the IServiceProvider
    // is available.
    public interface IDeferredTracerProviderBuilder
    {
        IServiceCollection Services { get; }

        TracerProviderBuilder Configure(Action<IServiceProvider, TracerProviderBuilder> configure);
    }
```

### Hosting

```csharp
    // These extensions enable IServiceProvider registration scenarios
    public static class TracerProviderBuilderExtensions
    {
        public static TracerProviderBuilder AddInstrumentation<T>(this TracerProviderBuilder tracerProviderBuilder)
            where T : class

        public static TracerProviderBuilder AddProcessor<T>(this TracerProviderBuilder tracerProviderBuilder)
            where T : BaseProcessor<Activity>

        public static TracerProviderBuilder SetSampler<T>(this TracerProviderBuilder tracerProviderBuilder)
            where T : Sampler

        public static TracerProviderBuilder Configure(this TracerProviderBuilder tracerProviderBuilder, Action<IServiceProvider, TracerProviderBuilder> configure)

        public static TracerProvider Build(this TracerProviderBuilder tracerProviderBuilder)
    }
```

## Usage Examples

### Binding exporter options to IConfiguration (#101, #343):

```csharp
    services.AddOpenTelemetryTracing((builder) => builder
        .AddJaegerExporter());
    services.Configure<JaegerExporterOptions>(this.Configuration.GetSection("Jaeger"));
```

### Registering a Processor using DI:

```csharp
    services.AddOpenTelemetryTracing((builder) => builder
        .AddProcessor<MyProcessor>());

    private class MyProcessor : BaseProcessor<Activity>
    {
        private readonly ILogger<MyProcessor> log;

        public MyProcessor(ILogger<MyProcessor> log)
        {
            this.log = log ?? throw new ArgumentNullException(nameof(log));
        }
    }
```

### Extension method that also registers services (#894):

```csharp
    services.AddOpenTelemetryTracing((builder) => builder
        .AddMyProcessor());

    public static class MyProcessorExtensions
    {
        public static TracerProviderBuilder AddMyProcessor(this TracerProviderBuilder tracerProviderBuilder)
        {
            if (tracerProviderBuilder is IDeferredTracerProviderBuilder deferredTracerProviderBuilder)
            {
                deferredTracerProviderBuilder.Services.AddHostedService<MyHostedService>();
                deferredTracerProviderBuilder.Services.AddSingleton<MyService>();
                deferredTracerProviderBuilder.AddProcessor<MyProcessor>();
            }

            return tracerProviderBuilder;
        }
    }
```

## Implementation Examples

### This is how JaegerExporter can use dependency injection to get `IOptions<JaegerExporterOptions>` from the IServiceProvider (if the application is using it).

```csharp
    public static TracerProviderBuilder AddJaegerExporter(this TracerProviderBuilder builder, Action<JaegerExporterOptions> configure = null)
    {
        if (builder == null)
        {
            throw new ArgumentNullException(nameof(builder));
        }

        if (builder is IDeferredTracerProviderBuilder deferredTracerProviderBuilder)
        {
            return deferredTracerProviderBuilder.Configure((sp, builder) =>
            {
                AddJaegerExporter(builder, sp.GetOptions<JaegerExporterOptions>(), configure);
            });
        }

        return AddJaegerExporter(builder, new JaegerExporterOptions(), configure);
    }

    private static TracerProviderBuilder AddJaegerExporter(TracerProviderBuilder builder, JaegerExporterOptions options, Action<JaegerExporterOptions> configure = null)
    {
        configure?.Invoke(options);

        var jaegerExporter = new JaegerExporter(options);

        if (options.ExportProcessorType == ExportProcessorType.Simple)
        {
            return builder.AddProcessor(new SimpleActivityExportProcessor(jaegerExporter));
        }
        else
        {
            return builder.AddProcessor(new BatchActivityExportProcessor(
                jaegerExporter,
                options.BatchExportProcessorOptions.MaxQueueSize,
                options.BatchExportProcessorOptions.ScheduledDelayMilliseconds,
                options.BatchExportProcessorOptions.ExporterTimeoutMilliseconds,
                options.BatchExportProcessorOptions.MaxExportBatchSize));
        }
    }
```

CodeBlanch commented on src/OpenTelemetry.Exporter.Jaeger/JaegerExporterHelperExtensions.cs:
```diff
@@ -38,7 +38,16 @@ public static TracerProviderBuilder AddJaegerExporter(this TracerProviderBuilder
 
-            var exporterOptions = new JaegerExporterOptions();
+            JaegerExporterOptions exporterOptions;
+            if (builder is IResolvingTracerProviderBuilder resolvingTracerProviderBuilder)
```
Here is an example of how something like JaegerExporter can use IResolvingTracerProviderBuilder to inject options.

reyang replied:
Reminded me of this https://github.com/census-instrumentation/opencensus-python/issues/394#issuecomment-482772410.

CodeBlanch replied:
@reyang Is that a good thing or a bad thing? I'm hoping since it was your comment it is a good thing 🤣

reyang replied:
It is a very good thing, most of the customers don't want to learn how the engine works, they just want to drive the car.
These configuration is the steering wheel.

CodeBlanch commented on examples/AspNetCore/Startup.cs:
```diff
@@ -63,20 +63,17 @@ public void ConfigureServices(IServiceCollection services)
-                        }));
+                        .AddJaegerExporter());
+
+                    services.Configure<JaegerExporterOptions>(this.Configuration.GetSection("Jaeger"));
```
Example of JaegerExporterOptions being configured using IOptions.

CodeBlanch commented:
@ejsmith 
> Looks interesting. Not seeing how I would go about registering services (IServicesCollection) during the build phase. I will check this branch out and take a look.

Ya, you're right, it only has the IServiceProvider. I'll keep messing with it and see if I can make that work too. Feel free to ping me on slack if you have any ideas. Or competing PR is also welcome. I feel like this is possible but it might take a couple of tries 😄

CodeBlanch commented:
@ejsmith I just pushed a new extension method called `ConfigureOpenTelemetryTracing` for your use case. Let me know what you think about that.

`TracerProviderBuilder` doesn't exist until `IServiceProvider` is ready. I think the pattern you are looking for is the builder lives during the `IServiceCollection` phase and then after `Build` you can no longer modify the services. I don't disagree with you, but all of the existing methods "adding" or "setting" things on the builder would need to be reworked to make that happen. What this new `ConfigureOpenTelemetryTracing` extension method does is allow you configure `IServiceCollection` and then register a callback for the "Build" phase.

Example...

```csharp
services.AddOpenTelemetryTracing((builder) => builder
   .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService(this.Configuration.GetValue<string>("Jaeger:ServiceName")));
   .AddJaegerExporter());
services.Configure<JaegerExporterOptions>(this.Configuration.GetSection("Jaeger"));
services.AddOpenTelemetryEventLogging(); // Pretend this is the new thing from your PR. 
```

`AddOpenTelemetryEventLogging` extension looks something like this...

```csharp
public static IServiceCollection AddOpenTelemetryEventLogging(this IServiceCollection services) {
   services.AddHostedService<MyHostedService>();
   services.AddSingleton<MyService>();
   services.ConfigureOpenTelemetryTracing((sp, builder) => builder.AddProcessor<MyProcessor>()); // This will be called with the other stuff from above when the TracerProvider is being built up.
}
```

I know it isn't exactly what you were looking for, but it would work?

ejsmith commented on src/OpenTelemetry/Trace/IResolvingTracerProviderBuilder.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace OpenTelemetry.Trace
+{
+    public interface IResolvingTracerProviderBuilder
```
I don't like this because you are limiting yourself to these 2 methods for resolving services instead of just using the normal `IServiceProvider` with all the functionality that everyone is used to in this .NET world. `IServiceProvider` is in the base .NET code and has been forever. So you don't have to take any dependencies. I'd recommend that this interface just add a property for `IServiceProvider`.

CodeBlanch replied:
Thanks for the tip! Didn't realize (or had forgotten) `IServiceProvider` lives in System. I pushed an update so `IResolvingTracerProviderBuilder : IServiceProvider`. I didn't drop `IResolvingTracerProviderBuilder` completely because of the `IOptions<T>` case. In order to get options, you have to ask for IOptions or IOptionsMonitor or IOptionsSnapshot which is going to painful in exporters that might not have the dependency on options. So at the moment `IResolvingTracerProviderBuilder` exists to expose a helper to get those. Thoughts?

ejsmith replied:
Hmm... don't really like that the builder itself is implementing `IServiceProvider` instead of just exposing a property that is an `IServiceProvider`. Not really understanding the options thing. I'd like to experiment with the code, but haven't really gotten a chance to yet.

ejsmith commented on src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs:
```diff
@@ -24,7 +24,7 @@ namespace OpenTelemetry.Trace
     /// Build TracerProvider with Resource, Sampler, Processors and Instrumentation.
     /// </summary>
-    internal class TracerProviderBuilderSdk : TracerProviderBuilder
+    public class TracerProviderBuilderSdk : TracerProviderBuilder
```
This is good and something that I think opens a lot of possibilities. I think what would be even better is to change up `TracerProviderSdk` to take factories for the processors just like it already does for the instrumentations. Should probably use use a factory for the sampler. If we made those 2 changes then we could actually split this all up into the config and the build phase and we could use the factory methods in the hosting version to call into the IServiceProvider to get those instances.

ejsmith replied:
Also, in general, there are too many parameters going into `TracerProviderSdk`. Would be better to create an options object that has all this stuff so that new things can be added without adding more and more parameters to the `TracerProviderSdk` ctor.

CodeBlanch replied:
Let me take a look and see what I can cook up. I think most of the stuff is `internal` in there so we should be able to change without breaking things.

CodeBlanch replied:
OK a new version has been pushed. I'm going to add some examples to the description.

ejsmith commented:
I created a PR to show what I am thinking for being able to separate the configuration and construction phases which I think would help with this PR.

ejsmith commented on src/OpenTelemetry.Exporter.Zipkin/ZipkinExporterHelperExtensions.cs:
```diff
@@ -38,22 +38,35 @@ public static TracerProviderBuilder AddZipkinExporter(this TracerProviderBuilder
-            var exporterOptions = new ZipkinExporterOptions();
-            configure?.Invoke(exporterOptions);
-            var zipkinExporter = new ZipkinExporter(exporterOptions);
+            if (builder is IDeferredTracerBuilder deferredTracerBuilder)
+            {
+                return deferredTracerBuilder.Configure((sp, builder) =>
+                {
+                    AddZipkinExporter(builder, sp.GetOptions<ZipkinExporterOptions>(), configure);
+                });
+            }
+
+            return AddZipkinExporter(builder, new ZipkinExporterOptions(), configure);
```
I personally don't like how this approach is causing all exporters and instrumentations to have to type check the builder and then have different code paths. IMO, it shouldn't be optional for these exporters and instrumentations to support separating configuration and construction phases. It should be mandatory that they register things with a deferred approach so that we can keep configuration and construction phases cleanly separated.

ejsmith replied:
I would personally prefer if we could force all of them to be registered with lambdas and not even make it an option to immediately new them up when configuring.

CodeBlanch replied:
Of everything on this PR, I'm the least happy with this bit. But I don't think it is possible to do it any other way 😭 It's not really a problem of separating configuration and construction, this is about the inversion of dependencies we've created for ourselves. To do it "properly" we would need this ctor: `public ZipkinExporter(IOptions<ZipkinExporterOptions> options)`. But that of course requires a direct dependency.

ejsmith commented on src/OpenTelemetry.Extensions.Hosting/Implementation/TracerProviderBuilderHosting.cs:
```diff
@@ -0,0 +1,123 @@
+    {
+        private readonly List<InstrumentationFactory> instrumentationFactories = new List<InstrumentationFactory>();
+        private readonly List<Type> processorTypes = new List<Type>();
+        private readonly List<Action<IServiceProvider, TracerProviderBuilder>> configurationActions = new List<Action<IServiceProvider, TracerProviderBuilder>>();
```
I like having generic configuration actions like this.

ejsmith commented on src/OpenTelemetry.Extensions.Hosting/Trace/TracerProviderBuilderExtensions.cs:
```diff
@@ -0,0 +1,67 @@
+        {
+            if (tracerProviderBuilder is TracerProviderBuilderHosting tracerProviderBuilderHosting)
+            {
+                tracerProviderBuilderHosting.AddInstrumentation<T>();
+            }
+
+            return tracerProviderBuilder;
+        }
+
+        public static TracerProviderBuilder AddProcessor<T>(this TracerProviderBuilder tracerProviderBuilder)
+            where T : BaseProcessor<Activity>
+        {
+            if (tracerProviderBuilder is TracerProviderBuilderHosting tracerProviderBuilderHosting)
+            {
+                tracerProviderBuilderHosting.AddProcessor<T>();
+            }
+
+            return tracerProviderBuilder;
+        }
+
+        public static TracerProviderBuilder SetSampler<T>(this TracerProviderBuilder tracerProviderBuilder)
+            where T : Sampler
+        {
+            if (tracerProviderBuilder is TracerProviderBuilderHosting tracerProviderBuilderHosting)
+            {
+                tracerProviderBuilderHosting.SetSampler<T>();
+            }
+
+            return tracerProviderBuilder;
+        }
+
+        public static TracerProviderBuilder Configure(this TracerProviderBuilder tracerProviderBuilder, Action<IServiceProvider, TracerProviderBuilder> configure)
+        {
+            if (tracerProviderBuilder is IDeferredTracerBuilder deferredTracerBuilder)
+            {
+                deferredTracerBuilder.Configure(configure);
+            }
+
+            return tracerProviderBuilder;
+        }
```
All of these duplicate extension methods wouldn't be necessary if the core SDK builder just used lambda factory methods for configuration.

ejsmith commented on src/OpenTelemetry/ServiceProviderExtensions.cs:
```diff
@@ -0,0 +1,28 @@
+    {
+        private static readonly Type OptionsGeneircType = Type.GetType("Microsoft.Extensions.Options.IOptions`1, Microsoft.Extensions.Options");
+
+        public static T GetOptions<T>(this IServiceProvider serviceProvider)
+            where T : class, new()
+        {
+            if (OptionsGeneircType == null)
+            {
+                throw new InvalidOperationException("Microsoft.Extensions.Options.IOptions<> could not be found reflectively in the running process.");
+            }
+
+            Type optionsType = OptionsGeneircType.MakeGenericType(typeof(T));
+
+            object options = serviceProvider.GetService(optionsType);
+            if (options == null)
+            {
+                return new T();
+            }
+
+            return (T)optionsType.GetProperty("Value").GetMethod.Invoke(options, null);
+        }
```
Again, we wouldn't need this if the core SDK supported factory methods for configuration because you could just register your exporter on that platform with a factory method that used the native API to get configuration options when new'ing up the exporter or instrumentation instance.

CodeBlanch commented:
@ejsmith I'll leave some comments on #1901.

CodeBlanch commented:
Pushed some updates:

* Switched to `public abstract class TracerProviderBuilderBase` so you can't new an instance. Previously version you could do `new TracerProviderBuilderSdk` but we want construction to happen through the static methods.
* Since SDK already has a dependency on IOptions, I removed the reflection stuff. Made the IServiceProvider extensions internal.

CodeBlanch commented:
@alanwest @reyang @cijothomas @ejsmith I incorporated all the feedback I got from people (on PR and on Slack) please take a look and let me know if this is something worth completing (unit tests, documentation, etc.). I tried to make it as clean as I could with the smallest public API footprint possible while still providing the use cases and without forcing any users into IServiceProvider.

ejsmith commented:
This is looking pretty good @CodeBlanch. I am going to try and play with the code soon, but I think it's a good approach.

reyang commented on src/OpenTelemetry/Trace/IDeferredTracerBuilder.cs:
```diff
@@ -0,0 +1,33 @@
+namespace OpenTelemetry.Trace
+{
+    /// <summary>
+    /// Describes a tracer builder tha support deferred initialization using a <see cref="IServiceProvider"/> to perform dependency injection.
```
```suggestion
    /// Describes a tracer builder that support deferred initialization using a <see cref="IServiceProvider"/> to perform dependency injection.
```

CodeBlanch replied:
Updated. Not sure what happened there 🤣

reyang replied:
Worked too hard without rest :)

reyang commented on src/OpenTelemetry/Trace/IDeferredTracerBuilder.cs:
```diff
@@ -0,0 +1,33 @@
+    /// <summary>
+    /// Describes a tracer builder tha support deferred initialization using a <see cref="IServiceProvider"/> to perform dependency injection.
+    /// </summary>
+    public interface IDeferredTracerBuilder
```
IDeferredTracer**Provider**Builder?

CodeBlanch replied:
Good call, updated.

reyang commented:
> @alanwest @reyang @cijothomas @ejsmith I incorporated all the feedback I got from people (on PR and on Slack) please take a look and let me know if this is something worth completing (unit tests, documentation, etc.). I tried to make it as clean as I could with the smallest public API footprint possible while still providing the use cases and without forcing any users into IServiceProvider.

The approach looks good to me 👍.
Instead of adding more to this PR (which is already a big one), I suggest that we put unit test / documentation in a follow up PR.

CodeBlanch commented:
> Instead of adding more to this PR (which is already a big one), I suggest that we put unit test / documentation in a follow up PR.

Sounds good to me. There are a couple of broken tests I'll need to fix up. I'll do that but I'm going to leave this as a draft until @ejsmith has a chance to test out his options approach. We can go with whichever people like more.

ejsmith commented on src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs:
```diff
@@ -99,19 +102,29 @@ public static TracerProviderBuilder AddProcessor(this TracerProviderBuilder trac
-            if (tracerProviderBuilder is TracerProviderBuilderSdk tracerProviderBuilderSdk)
+            if (tracerProviderBuilder is IDeferredTracerProviderBuilder)
+            {
+                throw new NotSupportedException("DeferredTracerBuilder requires a ServiceProvider to build.");
```
This can't be here because the `Build` method from the `TracerProviderBuilderHosting` class calls this method and then it blows up the build process.

CodeBlanch replied:
Whoops! Fixed it up by making the build on TracerProviderBuilderBase protected. I feel like we should keep this throw here to prevent anyone from accidentally using the hosting builder with the non-IServerProvider build extension?

ejsmith commented:
> > Instead of adding more to this PR (which is already a big one), I suggest that we put unit test / documentation in a follow up PR.
> 
> Sounds good to me. There are a couple of broken tests I'll need to fix up. I'll do that but I'm going to leave this as a draft until @ejsmith has a chance to test out his options approach. We can go with whichever people like more.

This approach is good with me @CodeBlanch. I experimented with a pretty similar approach after you said you didn't really like the options being public. I also realized that we need need to use the builder immediately and we can have a hosting builder that lives by itself which is what you are doing. I think there are some flaws in the current public API that limit things, but I know that we can't change those now after it was marked stable with 1.0. So I think your approach is the best one.
